solutions/usaco-1211.mdx

```diff
@@ -0,0 +1,100 @@
+---
+id: usaco-1211
+source: USACO Gold 2022 February
+title: Moo Network
+author: Jeffrey Zhang
+---
+
+[Official Analysis (Java)](http://usaco.org/current/data/sol_prob3_gold_feb22.html)
+
+## Solution
+
+The subtask of this problem can be solved using brute force: insert all the edges from $N$ points into an edge list and solve using [Kruskal's MST algorithm](https://usaco.guide/gold/mst#kruskals). Upon closer look at our brute force solution, the actual bottleneck is the time to insert all the edges.
+So our focus should be on finding a way to exploit the constraint $0 \le y_i \le 10$, so that the edges we need to insert can be reduced.
+ 
+Given three points $p_1$, $p_2$, and $p_3$, with coordinates $(x_1, y_1)$, $(x_2, y_2)$, and $(x_3, y_3)$ respectively, where $x_1 < x_2 < x_3$. We observe that if $y_1 = y_2$, the length of $p_1 p_3$ will always be larger than $p_1 p_2$ and $p_2 p_3$ individually. According to the nature of Kruskal's MST algorithm, edges being considered are sorted from shortest to longest, and the two pending points are in separate connected components. The $p_1 p_3$ edge being the longest edge will always be behind $p_1 p_2$ and $p_2 p_3$, and when its turn comes, $p_1$ and $p_3$ are already connected. So we can see that $p_1 p_3$ is an extraneous edge that we can safely remove.
+ 
+We can systematically ignore $p_1 p_3$ edges by sweeplining the points from left to right while keeping an array of points with the highest $x$ value in their $y$ lane. With every newer point, we will only be inserting edges with points from this array. This approach allows us to reduce the time complexity for inserting edges from $\mathcal{O}(N^2)$ to $\mathcal{O}(N\text{max}(y_i))$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}((N\text{max}(y_i))(\text{log}(\text{max}(y_i))+\text{log}(N)))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+typedef long long ll;
+
+//BeginCodeSnip{DSU}
+ll parent(vector<int>& top, int a){
+	if (top[a] == a) {
+		return a;
+	}
+	return top[a] = parent(top, top[a]);
+}
+
+bool merge(vector<int>& top, vector<int>& sizes, int a, int b){
+	a = parent(top, a);
+	b = parent(top, b);
+	if (a == b) {
+		return false;
+	}
+	if (sizes[b] > sizes[a]) {
+		swap(a,b);
+	}
+	top[b] = a;
+	sizes[a] += sizes[b];
+	return true;
+}
+//EndCodeSnip
+
+int main() {
+	int n;
+	cin >> n;
+	// Each array stores the point's {x, y, index} values
+	vector<array<ll, 3>> points(n);
+	for (int i = 0; i < n; i++) {
+		cin >> points[i][0] >> points[i][1];
+		points[i][2] = i;
+	}
+	sort(points.begin(), points.end());
+
+	// Buffer retains the highest x value point for a given y value
+	vector<array<ll, 3>> buffer(11, {-1, -1, -1});
+	// Edges stores a list of weighted edges
+	vector<array<ll, 3>> edges;
+	for (int i = 0; i < n; i++) {
+		for (int j = 0; j <= 10; j++) {
```

Comment:
what does this `10` mean

Reply:
I wrote 10 here because according to the constraints given by the problem, the y coordinate of any input point can only exist between 0 and 10